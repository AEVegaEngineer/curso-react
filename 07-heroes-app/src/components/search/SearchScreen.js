import React from 'react'
import queryString from 'query-string'
import { useLocation } from 'react-router-dom';
import { heroes } from '../../data/heros'
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';

export const SearchScreen = ({ history }) => {

  const location = useLocation();
  // {q = ''} cachea si viene undefined y retorna ''
  const {q = ''} = queryString.parse(location.search)
  console.log(q);
  
  const heroesFiltered = heroes;

  // si refresco el navegador, el valor de la caja de texto va a ser el que se recibe en la query del url
  const initialForm = {
    search: q
  }
  const [formValues, handleInputChange] = useForm(initialForm);
  const {search} = formValues;

  const handleSearch = (e) => {
    e.preventDefault();
    history.push(`?q=${search}`);
  }  

  return (
    <div>
      <h1>Search Screen</h1>
      <hr/>
      <div className="row">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr/>
          <form onSubmit={handleSearch}>
            <input 
              type="text" 
              className="form-control" 
              placeholder="Find your hero" 
              name="search" 
              autoComplete="off"
              value={search}
              onChange={handleInputChange}
            />
            <div className="d-grid gap-2">
              <button 
                className="btn btn-outline-primary btn-block" 
                type="submit"
              >
                Search...
              </button>
            </div>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr/>
            {
              heroesFiltered.map(hero => (
                <HeroCard key={hero.id} {...hero}/>
              ))
            }
        </div>
      </div>
    </div>
  )
}
