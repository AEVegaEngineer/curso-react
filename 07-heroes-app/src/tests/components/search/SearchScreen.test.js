import React from 'react'
import { mount } from 'enzyme'
import { MemoryRouter, Route } from 'react-router';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('pruebas en <SearchScreen/>', () => {
  test('debe de mostrarse correctament con valores por defecto', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path="/search" component={SearchScreen}/>
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search a Hero');
  });

  test('debe de mostrar a batman y el input con el queryString', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <Route path="/search" component={SearchScreen}/>
      </MemoryRouter>
    );
    expect(wrapper.find('input').prop('value').trim()).toBe('batman');
    expect(wrapper).toMatchSnapshot();
  });
  test('debe de mostrar un error si no se encuentra el Hero', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <Route path="/search" component={SearchScreen}/>
      </MemoryRouter>
    );
    expect(wrapper.find('.alert-danger').text().trim()).toBe("There is no hero named batman123");
    
  });
  
  test('debe de llamar el push del history', () => {
    const historyMock = {      
      push: jest.fn(),
    }

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <Route 
          path="/search" 
          // de esta manera no puedo enviar el argumento history
          //component={SearchScreen}
          // se debe hacer asi
          component={() => <SearchScreen history={historyMock}/>}
        />
      </MemoryRouter>
    );
    wrapper.find('input').simulate('change',{ 
      target : { 
        name: 'searchText',
        value: 'batman'
      }
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault(){}
    });

    expect(historyMock.push).toHaveBeenCalledWith(`?q=batman`)
  });
  
  
  
  
});