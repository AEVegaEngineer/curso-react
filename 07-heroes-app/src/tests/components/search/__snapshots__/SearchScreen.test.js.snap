// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <SearchScreen/> debe de mostrar a batman y el input con el queryString 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-outline-primary btn-block"
            type="submit"
          >
            Search...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card"
        style={
          Object {
            "maxWidth": "18rem",
            "minWidth": "18rem",
            "width": "18rem",
          }
        }
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                More...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`pruebas en <SearchScreen/> debe de mostrarse correctament con valores por defecto 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-outline-primary btn-block"
            type="submit"
          >
            Search...
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero
      </div>
    </div>
  </div>
</div>
`;
